Source/Programs/wavm.cpp

```diff
@@ -0,0 +1,224 @@
+			outputASTFilename = *args;
+		}
+		else if(!strcmp(*args, "--check") || !strcmp(*args, "-c"))
+		{
+			onlyValidate = true;
+		}
+		else if(!strcmp(*args, "--"))
+		{
+			++args;
+			break;
+		}
+		else if(!strcmp(*args, "--help") || !strcmp(*args, "-h"))
+		{
+			showHelp();
+			return EXIT_SUCCESS;
+		}
+		else {
+			break;
+		}
+	}
+
+	AST::Module* module = nullptr;
+	const char* main_arg0 = 0;
+	if (sourceFile)
+	{
+		WebAssemblyText::File wastFile;
+		if(loadTextModule(sourceFile,wastFile)) { module = wastFile.modules[0]; }
+		else { return EXIT_FAILURE; }
+		main_arg0 = sourceFile;
+	}
+	else if (binaryFile)
+	{
+		module = loadBinaryModule(binaryFile,binaryMemFile);
+		main_arg0 = binaryFile;
+	}
+	else
+	{
+		showHelp();
+		return EXIT_FAILURE;
+	}
+
+	if(!module) { return EXIT_FAILURE; }
+	if(!Runtime::init()) { return EXIT_FAILURE; }
+	if(!Runtime::loadModule(module)) { return EXIT_FAILURE; }
+
+	auto functionExport = module->exportNameToFunctionIndexMap.find(functionName);
+	// Support both main and _main by default
+	if(functionExport == module->exportNameToFunctionIndexMap.end() && !strncmp(functionName, "main", 4)) { functionExport = module->exportNameToFunctionIndexMap.find("_main"); }
+
+	if(functionExport == module->exportNameToFunctionIndexMap.end())
+	{
+		std::cerr << "Module does not export '" << functionName << "'" << std::endl;
+		return EXIT_FAILURE;
+	}
+
+	if(onlyValidate) { return EXIT_SUCCESS; }
+
+	if(outputASTFilename)
+	{
+		#if WAVM_TIMER_OUTPUT
+		Core::Timer printTimer;
+		#endif
+		std::ofstream outputStream(outputASTFilename);
+		if(!outputStream.is_open())
+		{
+			std::cerr << "Failed to open " << outputASTFilename << ": " << std::strerror(errno) << std::endl;
+			return EXIT_FAILURE;
+		}
+		outputStream << WebAssemblyText::print(module);
+		outputStream.close();
+		#if WAVM_TIMER_OUTPUT
+		printTimer.stop();
+		std::cout << "Printed WAST code in " << printTimer.getMilliseconds() << "ms" << std::endl;
+		#endif
+		return EXIT_SUCCESS;
+	}
+
+	Runtime::Value *parameters = nullptr;
+	auto function = module->functions[functionExport->second];
+	if(!strcmp(functionName, "main") || !strcmp(functionName, "_main"))
+	{
+		if (function->type.parameters.size() == 2)
+		{
+			uint32 main_argc_start = args-argv-1;
+			auto main_argc = (uint32)argc - main_argc_start;
+			auto main_argv = (uint32)Runtime::vmGrowMemory((main_argc+1) * sizeof(uint32));
+			auto main_argvGlobal = &Runtime::instanceMemoryRef<uint32>(main_argv);
+			for (auto i = main_argc_start; i < (uint32)argc; ++i)
+			{
+				const char* str = argv[i];
+				if (i == main_argc_start) { str = main_arg0; }
+				auto len = strlen(str)+1;
+				auto vmAddress = (uint32)Runtime::vmGrowMemory(sizeof(uint8)*(len));
+				memcpy(&Runtime::instanceMemoryRef<short>(vmAddress),str,len);
+				main_argvGlobal[i-main_argc_start] = vmAddress;
+			}
+			main_argvGlobal[main_argc] = 0;
+			Runtime::Value mainParameters[2] = {main_argc, main_argv};
```

Comment:
This is a bug: it sets `parameters` to point to this array, and uses the pointer after the array goes out of scope. IMO just make parameters a `std::vector`.